Add Undefined Behavior sanitizer to Swift Driver

rahul-malik opened the pull request:
## What's in this pull request?

This PR adds the ability to use the undefined behavior sanitizer. It accomplishes this by adding `--sanitize=undefined` to the Swift driver and linking the runtime library (if available) when it is specified.

This change is a dependency of adding this feature to SPM as discussed in [this thread](https://forums.swift.org/t/support-for-alternate-linkers-for-darwin-toolchain-via-use-ld/15008/8)

+ @jrose-apple @cheshire

rahul-malik committed:
Add Undefined Behavior sanitizer to Swift Driver

This change allows the swift driver to link the ubsan runtime if
`-sanitize=undefined` is specified.

jrose-apple commented on test/Driver/sanitizers.swift:
```diff
@@ -20,6 +26,24 @@
+
+// RUN: %swiftc_driver -resource-dir %S/Inputs/fake-resource-dir/lib/swift/ -driver-print-jobs -sanitize=undefined -target x86_64-apple-macosx10.9 %s | %FileCheck -check-prefix=UBSAN -check-prefix=UBSAN_OSX %s
+// RUN: not %swiftc_driver -resource-dir %S/Inputs/fake-resource-dir/lib/swift/ -driver-print-jobs -sanitize=undefined -target x86-apple-macosx10.9 %s 2>&1 | %FileCheck -check-prefix=UBSAN_OSX_32 %s
+// RUN: not %swiftc_driver -resource-dir %S/Inputs/fake-resource-dir/lib/swift/ -driver-print-jobs -sanitize=undefined -target x86_64-apple-ios7.1 %s 2>&1 | %FileCheck -check-prefix=UBSAN_IOSSIM %s
```
Hm, this doesn't seem right. I'm pretty sure we support UBSan on all Apple platforms. @cheshire?

rahul-malik replied:
I think you’re right - I was largely basing these off the existing tsan tests and didn’t see the “not” prefix

rahul-malik replied:
Also one thing I’ve been trying to find an answer for is if ubsan is supported on Windows. If so I’ll update the windows toolchain as well

jrose-apple replied:
Looks like not yet. Even if we're wrong about that, we can always add it later.

rahul-malik replied:
Ok I've omitted it for now. Updated the tests, let me know if this looks right

rahul-malik commented on test/Driver/sanitizers.swift:
```diff
@@ -20,6 +26,24 @@
+// RUN: %swiftc_driver -resource-dir %S/Inputs/fake-resource-dir/lib/swift/ -driver-print-jobs -sanitize=undefined -target arm64-apple-tvos9.0 %s 2>&1 | %FileCheck -check-prefix=UBSAN_tvOS %s
+// RUN: %swiftc_driver -resource-dir %S/Inputs/fake-resource-dir/lib/swift/ -driver-print-jobs -sanitize=undefined -target i386-apple-watchos2.0 %s 2>&1 | %FileCheck -check-prefix=UBSAN_watchOS_SIM %s
+// RUN: %swiftc_driver -resource-dir %S/Inputs/fake-resource-dir/lib/swift/ -driver-print-jobs -sanitize=undefined -target armv7k-apple-watchos2.0 %s 2>&1  | %FileCheck -check-prefix=UBSAN_watchOS %s
+// RUN: not %swiftc_driver -resource-dir %S/Inputs/fake-resource-dir/lib/swift/ -driver-print-jobs -sanitize=undefined -target x86_64-unknown-windows-msvc %s 2>&1 | %FileCheck -check-prefix=UBSAN_WINDOWS %s
```
TODO: Confirm windows doesn't support ubsan

rahul-malik commented:
@jrose-apple - Updated the tests

jrose-apple commented on test/Driver/sanitizers.swift:
```diff
@@ -20,6 +26,24 @@
+ */
+
+// RUN: %swiftc_driver -resource-dir %S/Inputs/fake-resource-dir/lib/swift/ -driver-print-jobs -sanitize=undefined -target x86_64-apple-macosx10.9 %s | %FileCheck -check-prefix=UBSAN -check-prefix=UBSAN_OSX %s
+// RUN: %swiftc_driver -resource-dir %S/Inputs/fake-resource-dir/lib/swift/ -driver-print-jobs -sanitize=undefined -target x86-apple-macosx10.9 %s 2>&1 | %FileCheck -check-prefix=UBSAN_OSX_32 %s
```
The positive checks should be checking both `-check-prefix=UBSAN` and `-check-prefix=UBSAN_<OS>`.

rahul-malik replied:
Fixing! Is there a reason the Address Sanitizer Linux / Windows positive checks don't specify `-check-prefix=ASAN` ?

jrose-apple replied:
I suspect it was just a copy/paste mistake, like yours were.

jrose-apple replied:
Ah, wait, no, the `rpath` command is Apple-platform-specific. I guess it's okay to leave those as they were.

rahul-malik replied:
Makes sense, I'll address and run the test suite locally before updating

rahul-malik committed:
- Update positive tests for ubsan
 - Update positive asan tests that don't contain `-check-prefix=ASAN`
 - Add ubsan to windows toolchain

rahul-malik commented on test/Driver/sanitizers.swift:
```diff
@@ -64,5 +87,20 @@
+// UBSAN_watchOS_SIM: lib/swift/clang/lib/darwin/libclang_rt.ubsan_watchossim_dynamic.dylib
+// UBSAN_watchOS: lib/swift/clang/lib/darwin/libclang_rt.ubsan_watchos_dynamic.dylib
+// UBSAN_LINUX: lib/swift/clang/lib/linux/libclang_rt.ubsan-x86_64.a
+// UBSAN_WINDOWS: lib/swift/clang/lib/windows/clang_rt.ubsan_standalone-x86_64.lib
```
This lib name is based off @troughton earlier comment but I believe the logic in the windows driver will look for `clang_rt.ubsan-x86_64.lib` so this might require an update

jrose-apple replied:
Ah, you should be able to run the tests locally to find out! docs/Testing.md has information on how to run just the one test.

rahul-malik commented:
@jrose-apple @troughton - Thanks for the comments, updated!

rahul-malik committed:
Remove darwin only check from linux / windows test cases

jrose-apple commented:
@swift-ci Please test

rahul-malik commented:
@jrose-apple - I ran into the same test failure locally but it doesn't quite make sense to me why `-sanitize=undefined` is not found given that if i pass `-sanitize=undefined` to the `swiftc` executable used for testing it works.

The test error:
```
<unknown>:0: error: unsupported option '-sanitize=undefined' for target 'x86_64-apple-macosx10.9'
```

This error occurs if you pass an unsupported value for `-sanitize` or if we are unable to locate the dylib. I'm assuming this error is the latter but I am unsure how to debug this further without knowing the actual folder hierarchy and contents in the CI environment.

jrose-apple commented:
Maybe we just don't build the UBSan support by default. @cheshire, @Rostepher?

cheshire commented:
@rahul-malik Thanks! The source looks fine.
If the error is reproducible locally for you, could you try to build a debug version of the swift compiler, and use debugger to figure out what goes wrong?
IIRC the ubsan should be build by default.

The driver uses a somewhat convoluted logic to mark the sanitizer as unavailable if the file for it is not found in resource directory. Could you try putting a breakpoint there and seeing what happens?
If the file is not found, is the path wrong? In my build directory, I have a file "lib/clang/6.0.0/lib/darwin/libclang_rt.ubsan_osx.a" so it seems it should be found.

rahul-malik commented:
@Cheshire - I'll look into it! I noticed you referenced a static library (.a) but the source looks for a dylib. I'm assuming you have the dylib in the same folder as well?

In my local Xcode toolchain directory I have both available
```
❯ ls | grep ubsan
libclang_rt.ubsan_ios.a
libclang_rt.ubsan_ios_dynamic.dylib
libclang_rt.ubsan_iossim.a
libclang_rt.ubsan_iossim_dynamic.dylib
libclang_rt.ubsan_osx.a
libclang_rt.ubsan_osx_dynamic.dylib
libclang_rt.ubsan_tvos.a
libclang_rt.ubsan_tvos_dynamic.dylib
libclang_rt.ubsan_tvossim.a
libclang_rt.ubsan_tvossim_dynamic.dylib
libclang_rt.ubsan_watchos.a
libclang_rt.ubsan_watchos_dynamic.dylib
libclang_rt.ubsan_watchossim.a
libclang_rt.ubsan_watchossim_dynamic.dylib
```

The dylib is referenced from this method in DarwinToolchain.cpp
```

cheshire commented:
@rahul-malik yeah it's weird. What if you change the lookup to use the archive file, does everything work then?

rahul-malik commented:
@Cheshire - Thanks! Haven't been able to get back to this for a few days but will look into your suggestion tonight.

cheshire commented:
@rahul-malik Sorry, I am still somewhat confused on what exactly the problem is and to which degree is it isolated.

Is it correct to say that the test is failing, and you could not quite understand why?
If yes, then it could be a matter of just adding an empty (stub) file to `test/Driver/Inputs/fake-resource-dir/lib/swift/clang/lib`, alongside many others (e.g. check out `test/Driver/Inputs/fake-resource-dir/lib/swift/clang/lib/darwin`).
During the test, the flag `-resource-dir %S/Inputs/fake-resource-dir/lib/swift/` is passed to the driver, and that's why the look up is performed in that folder.

cheshire reviewed (changes requested):
Seems you need to add empty stub files for ubsan into `test/Driver/Inputs/fake-resource-dir`

rahul-malik commented:
Thank you! Yes I was unclear how to go about resolving the missing file stub issue. I’ll update the PR next week with this change

cheshire commented:
@rahul-malik NP! In general, I find that such issues are best resolved with debuggers: bisection narrows it down to the file not being found, and then debugger would let you know that it searches in a stub location.

rahul-malik committed:
Add test fake inputs, rebase latest swift master

rahul-malik committed:
Add remaining ubsan stubs

rahul-malik commented:
@cheshire @jrose-apple - Got around to updating with new inputs and rebasing the latest sources. Should be good to go 👍

cheshire commented:
@swift-ci please test

rahul-malik commented:
@jrose-apple @cheshire - Yay! Tests passed :) Is this PR ready to accept / merge?

cheshire commented:
OK, Thanks!

cheshire commented:
Should I merge it, or do you want to do it? You would need to use the "squash and merge" option.

jrose-apple commented:
Ah, external contributors don't have commit privileges by default, so you should be the one to merge it, George.
